crates/proof-of-sql/src/base/math/permutation.rs

```diff
@@ -84,6 +84,40 @@ impl Permutation {
             })
         }
     }
+
+    /// Apply the permutation to chunks of the given size within the slice
+    pub fn try_chunked_apply<T>(
+        &self,
+        slice: &[T],
+        chunk_size: usize,
+    ) -> Result<Vec<T>, PermutationError>
+    where
+        T: Clone,
+    {
+        if slice.len() % chunk_size != 0 {
+            return Err(PermutationError::PermutationSizeMismatch {
+                permutation_size: self.size(),
+                slice_length: slice.len(),
+            });
+        }
+
+        let num_chunks = slice.len() / chunk_size;
+        if self.size() != num_chunks {
+            return Err(PermutationError::PermutationSizeMismatch {
+                permutation_size: self.size(),
+                slice_length: num_chunks,
+            });
+        }
+
+        let mut result = Vec::with_capacity(slice.len());
+        for &i in &self.permutation {
+            let start = i * chunk_size;
+            let end = start + chunk_size;
+            result.extend_from_slice(&slice[start..end]);
+        }
+
+        Ok(result)
+    }
```

Comment:
This _may or not_ work but if it does it helps reduce the preallocation and yields a slightly more functional result:
```suggestion
pub fn try_chunked_apply<T>(
    &self,
    slice: &[T],
    chunk_size: usize,
) -> Result<Vec<T>, PermutationError>
where
    T: Clone,
{
    if slice.len() % chunk_size != 0 {
        return Err(PermutationError::PermutationSizeMismatch {
            permutation_size: self.size(),
            slice_length: slice.len(),
        });
    }
    let num_chunks = slice.len() 
    if self.size() != num_chunks {
        return Err(PermutationError::PermutationSizeMismatch {
            permutation_size: self.size(),
            slice_length: num_chunks,
        });
    }
    let result = self
        .permutation
        .iter()                           
        .flat_map(|&i| {
            slice[i * chunk_size .. (i + 1) * chunk_size]
                .iter()                    
        })
        .cloned()                        
        .collect();                      

    Ok(result)
}
```

Reply:
Thanks for the review. This works fine. just updated it